0011-chunking-o1js.md

```diff
@@ -0,0 +1,97 @@
+
+In order to make chunking work in o1js, we have to modify the `Pickles.compile` to accept the number of chunks that are required in order to construct the proof. The Pickles [test](https://github.com/MinaProtocol/mina/blob/develop/src/lib/pickles/test/chunked_circuits/test_chunked_circuits.ml) serves as an example on how to modify the `Pickles.compile` function to prove larger circuits and specify the amount of chunks.
+
+Calculating how many chunks are needed to prove a circuit depends on the amount of constraints used in a circuit. Before passing the parameter to the function and compiling the circuit, we have to calculate the amount of chunks needed. Luckily, o1js already has a function that can inspect methods and count the amount of constraints, we get this information by calling `.analyzeMethods()` on the `ZkProgram` or `SmartContract`.
+
+### Technical Implementation
+
+The implementation should not require and API changes for the developer, all changes happen internally and hidden from the public facing API. A naive overview of the changes:
+
+1. Adjusting OCaml bindings to Pickles - [pickles_bindings.ml](https://github.com/o1-labs/o1js-bindings/blob/1db0524c5c7c93f0df9db3fb3bdfa0414bdaf90b/ocaml/lib/pickles_bindings.ml#L563)
+
+   - Modifying the `pickles_compile` function to accept an additional parameter `num_chunks`, which will then be passed into `Pickles.compile_promise`. `num_chunks` will be calculates in o1js directly.
+
+2. Adapting the compile function in [proof_system.ts](https://github.com/o1-labs/o1js/blob/0d82ad1d67417c72b06fc8456a0a2b54bcc46652/src/lib/proof_system.ts)
+
+   - `compileProgram` function should accept a new argument `numChunks` which represents the amount of chunks needed to compile the circuit
+   - `numChunks` inside of `compileProgram` will then be passed into `Pickles.compile` via the bindings layer
+
+3. Before calling into `compileProgram`, the number of chunks must be calculated. This can be done by utilizing the `analyzeMethods` function which exists on both `SmartContract` and `ZkProgram`.
+   - For `ZkProgram`, this change can be done in `proof_systems.ts` and the `ZkProgram` function
+   - For `SmartContract`, the change is needed in [zkapp.ts](https://github.com/o1-labs/o1js/blob/0d82ad1d67417c72b06fc8456a0a2b54bcc46652/src/lib/zkapp.ts) - in the `SmartContract.compile` method.
+
+- ## Changing the type definitions of the bindings to represent the change in the `pickles_compile` function
+
+## Test plan and functional requirements
+
+1. Testing Goals and Objectives:
+
+- o1js can compile and prove circuits that are bigger than 2^16 constraints, up to ideally 2^32 (the new limit introduced by chunking)
+- A proof using a larger circuit limit can be deployed to and verified on a testnet
```

Comment:
By 'larger' do you mean > 2^16 and < 2^32? I assume so. Could also be interpreted as > 2^32. Might help to make this explicit